write client with typescript

fabienjuif opened the pull request:
fixes #16

## What

- adding a `.npmignore` file
- adding a prettier config file
- adding a `format` script that format `src/` and `tests/` directories thanks to prettier
- running prettier over all files
- adding a `build` script that compile `src/` into `build/` thanks to tsc
- create first types by retroengineering responses (`src/types/`)
- create first typescript version of the Tado class (`src/index.ts`)
- export the Tado class as a named class instead of using `default`
- settings the `types` entry into the `package.json`
- adding myself as contributor
- `updatePresence` use Promise.all


Note that I choose to push the build/ directory if some folks use the git repository as a dependency instead of npm.

## TODOs
- [x] Finish typing the client
- [x] Maybe publish the fork on my own to see how it behaves

fabienjuif commented on src/index.ts:
```diff
@@ -0,0 +1,542 @@
+import { AccessToken, ResourceOwnerPassword } from 'simple-oauth2'
```
Here is your source code with some types added (I have to finish, help is welcome :))

fabienjuif commented on src/index.ts:
```diff
@@ -0,0 +1,542 @@
+        return response.data as R
+    }
+
+    getMe(): Promise<Me> {
```
You can see I removed the `async` key work to return the promise and avoid aving a nested Promise as the end (since we don't await)

fabienjuif commented on package.json:
```diff
@@ -1,36 +1,46 @@
+        "simple-oauth2": "^4.3.0"
+    },
+    "devDependencies": {
+        "@types/node": "^18.11.9",
+        "@types/simple-oauth2": "^4.1.1",
```
These are new

fabienjuif commented on package.json:
```diff
@@ -1,36 +1,46 @@
+        "mocha": "^10.0.0",
+        "nock": "^13.2.0",
+        "nyc": "^15.1.0",
+        "prettier": "^2.7.1",
+        "typescript": "^4.8.4"
```
These are new

mattdavis90 commented:
Hi @fabienjuif I wanted you to know that I've been keeping up with the commits over the last couple hours. I'm keen to see how this goes and if we can keep backward compatibility that's awesome! I'm happy to do some testing once you're finished with the changes. Give me an @ when you're ready to test and merge. Thanks

fabienjuif commented:
Hello @mattdavis90 :) 

This is **almost** backward compatible at the moment, here is what break things RN:
> export the Tado class as a named class instead of using default

I did it for DX (developer experience) because as far as I know, this is a pain to handle default exports in all worlds (commonjs/esm/typescript/etc)
If you really want to let it as a default export I can revert this!

FYI I keep all your test to make sure the change I do are okay + I have a client to test each exposed method with my own Tado [and to retrieive payload to type them], sadly I don't have a AC.

fabienjuif commented:
@mattdavis90 I can't do more with available data right now.

As you may notice there is two methods I can't call without the Tado API rejecting me: 
- getEnergyIQTariff
- updateEnergyIQTariff

I added // FIXME: on top of them

There are some type I can't guess.

And finally I extracted enums in a dedicated file so this is easier if someone want to help by adding a value I can't catch.

---

Feel free to modify the PR if needed.
If you want me to try to publish this version in a new name I'll do.

fabienjuif commented:
@mattdavis90 if you want to try it, you can install `@fabienjuif/node-tado-client` 
Related PR: https://github.com/fabienjuif/node-tado-client/pull/1

mattdavis90 commented:
Looks really good to me. I tried having a go with getting it to be fully backward compatible with `const Tado = require(...)` but couldn't get that working... that's ok though, we can probably put a note in the README and I can update the NodeRed library to use either use `import` or `Tado.Tado`

I'll have a go at calling the IQ functions and see if I can get type information for them.

Re: The tests I'd left in there. I'm glad they were helpful. I mostly left them as a reference for myself in case Tado change what the APIs respond with so really glad that's helped for typing. I don't have a Tado AC either, or any radiator thermostats to test so some of the samples are built from contributors.

mattdavis90 commented:
I get a 403 for the two IQ related functions - I'm not sure I've ever set it up in my app so I probably can't test that.

mattdavis90 commented:
Regarding the missing types

`getInstallations` returns me an empty list so I can't help type this
The parameters to the `setOverlay` functions relating to `fanLevel` and `swing` I _think_ are strings but are sort of enum-like - they were talked about [here](https://github.com/mattdavis90/node-red-contrib-tado-client/issues/40#issuecomment-978935900) Is there a TypeScript way to model enums?

fabienjuif commented:
Sur it will be enums like this one for exemple: https://github.com/mattdavis90/node-tado-client/pull/30/files#diff-c11b94ab4abb4e89a897c1aa5ab7db811cafb3a6853ba5825c6518d13ef25faeR15

mattdavis90 commented:
I'm just doing some testing with the NodeRed client to see what updates might be needed (if any) then I think I'm happy to merge :rocket:

mattdavis90 commented:
Seems to be working as long as the Node.Js version is >=14 - is that due to how the TypeScript was generated?

This isn't a major issue for me but there are people that use an old RasperryPi image for NodeRed and I've had bugs raised in the past because I switched to promises/async that needed an update from Node v10 to v12

This won't be a blocker IMO - just checking if there's any backward compatibility bits that could help people migrate. Thanks

fabienjuif commented:
Yeah I choose to transpile to ES2021 (or 2022) but we can target a version before that.
I though you were using a pretty recent NodeJS version because you used async/await without any transpilation.

fabienjuif commented:
Do you want me to change the target and publish a new version?

mattdavis90 commented:
That's working a treat on node v12. I personally run node v16 with NodeRed but I know from previous issues that others have never moved to newer versions and some of the RapsberryPi based distros haven't either. If there's no adverse affect to targeting ES2019 then this all looks good to me. Thanks for the hard work!

fabienjuif commented:
My previous commit lower the transpilation target and it work on node12 (but I just tested the login)

mattdavis90 commented:
Are you happy for me to merge and release?

fabienjuif commented:
Sure I am :)

mattdavis90 commented:
@fabienjuif Just to say a massive thank you for contributing this code! I added some new API calls today and it was a breeze because of the TS support! Have a very Happy Christmas and thanks again

fabienjuif commented:
You are welcome ❤️ 

I hope you had an Happy Christmas, and have a nice new eve!
